### PR TITLE
[FIX-IMP] sale(_stock): forecasted move by warehouse

### PR DESCRIPTION
**Pull request by ryv-odoo**
[IMP] (sale_)stock: forecast button with warehouse

In case of multi-warehouse, the forecast button won't
take in account the related warehouse and redirect to the
forecast report of first warehouse of the company.
To ease the understand of the forecast report redirect
with the right warehouse in the context.

[FIX] stock: bad warehouse forecast

In the forecast data computation of `stock.move`,
the warehouse wasn't take in account in the call of `_get_report_lines`
for the quantity available computation. Add the `location` key to the
context with all location related to the warehouse in
the `_get_report_lines`.

### CODE REVIEW

File: addons/stock/report/report_stock_forecasted.py [hunk 1]
```diff
@@ -143,7 +143,7 @@ def _get_report_lines(self, product_template_ids, product_variant_ids, wh_locati
         ins_per_product = defaultdict(lambda: [])
         for in_ in ins:
             ins_per_product[in_.product_id.id].append([in_.product_qty, in_])
-        currents = {c['id']: c['qty_available'] for c in outs.product_id.read(['qty_available'])}
+        currents = {c['id']: c['qty_available'] for c in outs.product_id.with_context(location=wh_location_ids).read(['qty_available'])}
```
Comment (Whenrow) [inline addons/stock/report/report_stock_forecasted.py:146]: there is the `warehouse` key available in `get_domain_locations()`

File: addons/stock/static/src/js/forecast_widget.js [hunk 2]
```diff
@@ -56,14 +56,14 @@ const ForecastWidgetField = AbstractField.extend({
...
         this._rpc({
-            model: resModel,
+            model: 'stock.move',
             method: 'action_product_forecast_report',
-            args: [productId],
+            args: [this.recordData.id],
         }).then(action => {
-            action.context = {
+            action.context = Object.assign(action.context || {}, {
                 active_model: resModel,
```
Comment (svs-odoo) [inline addons/stock/static/src/js/forecast_widget.js:64]: The `resModel` variable is useless if you use it only in one place. You can remove it and write directly `product.product` instead.

File: addons/stock/report/report_stock_forecasted.py (hunk 1 above)
Comment (svs-odoo) [inline addons/stock/report/report_stock_forecasted.py:146]: I'm not sure to understand why you need to do that ?
I though the `warehouse` put in the context in the `_get_report_data` was already enough to address to this commit's issue ?

Comment (ryv-odoo) [reply addons/stock/report/report_stock_forecasted.py:146]: It is not for the report it-self. But for the computation of forecasted information on the stock move which call `_get_report_lines` without set a warehouse in context.

File: addons/sale_stock/static/src/js/qty_at_date_widget.js [hunk 3]
```diff
@@ -80,6 +80,7 @@ var QtyAtDateWidget = Widget.extend({
         action.context = {
             active_model: 'product.product',
             active_id: this.data.product_id.data.id,
+            warehouse: this.data.warehouse_id && this.data.warehouse_id.res_id
```
Comment (amoyaux) [inline addons/sale_stock/static/src/js/qty_at_date_widget.js:83]: warehouse_id

**Comment by amoyaux** [comment]
robodoo r+

**Comment by ryv-odoo** [comment]
@robodoo rebase-ff

**Comment by amoyaux** [comment]
@robodoo rebase-ff

**Pull request closed without merging**